Integrate health checks into normal test execution

DRMacIver opened the pull request:
Previously most health checks ran as a separate phase in which the test body was not executed and the data generation was run in a weird way that didn't take into account the ways in which Hypothesis data generation is clever.

This has a number of disadvantages:

* It adds a bunch of extra time to the beginning of the run
* It doesn't actually agree with the details of how Hypothesis really generates data, so its measurements are all off - it will pass things that it should fail and vice versa
* It ignores things it shouldn't be ignoring such as data() and assume().

This fixes all that by instead integrating this part of the health check into the engine. We now keep track of the details of examples generated as use them to perform the same failure checks as we did before, but based on real test executions.

One thing that is important to note is where we do this. We don't do it from the first test that is run. Instead we do it after we have tried the trivial example (and thus, necessarily, after we have replayed any previous examples). This serves a roughly equivalent role to the way we previously generated and discard one example prior to running health checks, which potentially warms some caches to guard against the problem where running the first example can be slow and all subsequent ones are fine.

Zac-HD commented on src/hypothesis/internal/healthcheck.py:
```diff
@@ -0,0 +1,40 @@
+        '\nSee https://hypothesis.readthedocs.io/en/latest/health'
+        'checks.html for more information about this. '
+    )
+    message += (
```
Could add one string here instead of two.

Zac-HD commented:
This looks pretty good to me - I always forget `time.monotonic`, because I don't usually have a compat shim 😄 

My only question is whether this should affect the `phases` setting (I don't think so) or the `perform_health_check` setting (yes?).  Either case for the latter is unclear to me, so I'd appreciate hearing your thoughts!

DRMacIver commented:
> My only question is whether this should affect the phases setting (I don't think so)

I can't think of a reason why it should. There's no health check phase, and this just integrates with the existing phases

> or the `perform_health_check` setting (yes?).

No, the behaviour should be unchanged for `perform_health_check`. Note that we check it in the `begin_health_checks` method.

DRMacIver commented:
Ugh. This ended up necessitating a bunch of knock on changes. Our test suite has the weird property that a watched clock never boils (err or something like that) - the more you measure time, the slower it goes! So adding all those extra time measurements ended up slowing down the clock which caused a bunch of other tests to fail.

I've gone through and I've removed deadlines and health checks from a bunch of tests that need them. I *hope* I've got them all now, though there are probably still a few lurking. I'm sure CI will tell me.

One annoying change that I don't really understand is that `test_clears_out_everything_smaller_than_the_interesting_example` became flaky. I don't really understand why this should/could be the case - the change I made shouldn't have affected anything at all in that regard. However I didn't really feel like debugging it, so I replaced the test with a version of it that should be strictly better anyway.

DRMacIver commented:
OK. Now with test suite actually passing. Phew.

Zac-HD commented on tests/cover/test_conjecture_float_encoding.py:
```diff
@@ -109,13 +110,18 @@ def test_floats_round_trip(f):
 
 
-finite_floats = st.floats(allow_infinity=False, allow_nan=False, min_value=0.0)
+finite_floats = st.builds(
```
This is only used in the definition of `integral_floats` - I'd just combine them.

DRMacIver commented:
> I'd find this much easier to review if you could combine it into logical commits though, to separate the engine changes from tests.

Done.

Zac-HD commented on tests/cover/test_testdecorators.py:
```diff
@@ -203,7 +203,7 @@ def this_has_a_unique_name_and_lives_on_a_class(self, x):
 
-@given(lists(integers()), integers())
+@given(lists(integers(), unique=True), integers())
 def test_removing_an_element_from_a_unique_list(xs, y):
```
I don't follow what this is actually testing - it just seems both obviously correct and triply redundant.  Do we still have full coverage if it's deleted?

*Having read further, I think it's just needless contrast with the next test and should be deleted.*

DRMacIver replied:
The entirety of `test_testdecorators.py` is a bit weird and it's more of a smoke test module than anything very interesting in its own right. I'd rather just leave it alone as much as possible in this PR.

Zac-HD commented on tests/cover/test_testdecorators.py:
```diff
@@ -216,9 +216,9 @@ def test_removing_an_element_from_a_unique_list(xs, y):
-    assume(y in xs)
+@given(lists(integers(), average_size=25.0), data())
+def test_removing_an_element_from_a_non_unique_list(xs, data):
+    y = data.draw(sampled_from(xs))
```
I think the test could actually fail because `xs == []`, rather than anything about uniqueness.

We could also simplify the test - instead of drawing an integer sampled form the list and removing it, just `xs.pop()` because order isn't important here.

DRMacIver replied:
> I think the test could actually fail because xs == [], rather than anything about uniqueness.

No, that's fine.  `sampled_from([])` is just an empty strategy and drawing from it is equivalent to `reject()`.

Zac-HD commented on RELEASE.rst:
```diff
@@ -0,0 +1,27 @@
+improvements in accuracy. In particular:
+
+* New failures will usually be because they are now taking into account things
+  like use of :func:`~hypothesis.strategies.data()` and
```
nit: cross-references don't take parentheses - those are added by Sphinx.  I'm not sure if this would break the formatting, but I'd check that or just remove them.

Zac-HD commented on RELEASE.rst:
```diff
@@ -0,0 +1,27 @@
+RELEASE_TYPE: minor
+
+This release overhauls the health check system in a variety of small ways.
```
It might be nice to linkify this:

    :doc:`the health check system <healthchecks>`

Zac-HD commented on src/hypothesis/internal/conjecture/engine.py:
```diff
@@ -253,6 +276,74 @@ def test_function(self, data):
+                'Examples routinely exceeded the max allowable size. '
+                '(%d examples overran while generating %d valid ones)'
+                '. Generating examples this large will usually lead to'
+                ' bad results. You should try setting average_size or '
```
IIRC we no longer want to recommend using `average_size`; good time to take it out of the message?

Zac-HD commented on src/hypothesis/internal/conjecture/engine.py:
```diff
@@ -253,6 +276,74 @@ def test_function(self, data):
+                '%d valid examples in %.2f seconds (%d invalid ones '
+                'and %d exceeded maximum size). Try decreasing '
+                "size of the data you're generating (with e.g."
+                'average_size or max_leaves parameters).'
```
`average_size` -> `max_size`?

Zac-HD commented on src/hypothesis/internal/conjecture/engine.py:
```diff
@@ -641,6 +732,11 @@ def exit_with(self, reason):
         raise RunIsComplete()
 
+    def begin_health_checks(self):
+        if not self.settings.perform_health_check:
```
if self.settings.perform_health_check:
        self.health_check_state = HealthCheckState()

I think it might also be more readable as an inlined section with a brief explanatory comment - I would have had trouble following the effect of `self.health_check_state` if I hadn't been prepped by your writeup.
